src/replication.c

```diff
@@ -3841,11 +3912,24 @@ static int rdbChannelAbortRdbTransfer(void) {
     if (server.repl_rdb_ch_state == REPL_RDB_CH_STATE_NONE)
         return C_OK;
 
-    if (server.repl_rdb_transfer_s) {
-        if (server.loading) {
-            /* If loading flag is set, we want to handle cleanup asynchronously.
-             * We set REPL_RDB_CH_STATE_CLOSE_ASAP so it will be handled just
-             * outside loading loop.*/
+    /* This function may also be called if a problem is detected on the main
+     * channel. In this case, we handle the situation differently based on
+     * the current state:
+     * - If we started loading the RDB file and the RDB is disk-based, we mark
+     *   the main channel's state as CLOSE_ASAP and defer the failure handling
+     *   until after the RDB has been loaded. This way we allow the replica to
+     *   retry psync after the RDB is loaded.
+     * - For diskless loading, we cannot safely free the rdb channel connection
+     *   object. Instead, we mark the RIO object as aborted so the next
+     *   rioRead() will fail safely.
+     * - If the RDB has already been loaded, and we are streaming the
+     *   accumulated buffer to the database, we mark the main connection
+     *   as CLOSE_ASAP and wait until the accumulated buffer is drained.
```

Comment:
we will stop if finding the main channel is CLOSE_ASAP, right? why do we need to wait?

Reply:
We can stream whatever we have already accumulated, then replica can try psync. Streaming buffer is relatively fast, so we don't need to discard the data we have. I assumed that on a connection error, if we discard what we have, there is a chance that master needs to re-send these discarded data. Just to avoid that, made it that way, to increase the chance of successful psync with higher repl offset.

Reply:
thanks, i found we just only check `REPL_MAIN_CH_CLOSE_ASAP` at the beginning of  `rdbChannelStreamReplDataToDb` instead of during consuming the replication buffer, so we can support what you said above, do i understand correctly?

Reply:
hmm, wait, you are right. We need to move it after the loop.

Reply:
i think we just stop at the beginning, rdbChannelStreamReplDataToDb should be after rdb transferring

Reply:
oh, we don't stop just set `close_asap`

Reply:
See the last commit. I mean, before this commit, we wouldn't realize connection is broken immediately, it would wait until timeout. With the latest commit, if that flag is set, we'll free master client immediately.